### PR TITLE
Support arm64 in cpubuilder dockerfile using multi-arch builds.

### PR DESCRIPTION
**Pull request by ScottTodd**
This should let us replace https://github.com/iree-org/iree/blob/main/build_tools/docker/dockerfiles/base-arm64.Dockerfile with this cpubuilder dockerfile, making progress on https://github.com/iree-org/iree/issues/15332. It uses a multi-architecture build rather than a fully forked file, which seems to work reasonably well with some local testing. I can even run the arm64 dockerfile on my x86_64 host.

Various related changes are included here:

* Drop `x86_64` from file names
  * `publish_cpubuilder_x86_64.yml` --> `publish_cpubuilder.yml`
  * `cpubuilder_ubuntu_jammy_x86_64.Dockerfile` --> `cpubuilder_ubuntu_jammy.Dockerfile`
  * `cpubuilder_ubuntu_jammy_ghr_x86_64.Dockerfile` --> `cpubuilder_ubuntu_jammy_ghr.Dockerfile`
* Build with `--platform linux/amd64,linux/arm64` and update docs for this
* Build ccache from source as needed in `build_tools/install_ccache.sh` (code lifted from https://github.com/iree-org/iree/blob/main/build_tools/docker/context/install_ccache.sh). Note that if we standardize on sccache we can drop the ccache install entirely
  * Since this builds from source using CMake, I moved the ccache install step to after the cmake install step
* I am _not_ installing qemu here yet, as we did in IREE (https://github.com/iree-org/iree/blob/782f372b070eadd593a727004cf61dc84aabc634/build_tools/docker/dockerfiles/base-arm64.Dockerfile#L75-L80). I want to try installing on demand with https://github.com/docker/setup-qemu-action first.

### CODE REVIEW

File: dockerfiles/cpubuilder_ubuntu_jammy.Dockerfile [hunk 1]
```diff
@@ -15,32 +18,19 @@ RUN apt update && \
...
-RUN echo "deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-${LLVM_VERSION} main" >> /etc/apt/sources.list && \
-    curl https://apt.llvm.org/llvm-snapshot.gpg.key | gpg --dearmor > /etc/apt/trusted.gpg.d/llvm-snapshot.gpg && \
-    apt update && \
-    apt install -y clang-${LLVM_VERSION} lld-${LLVM_VERSION}
-# Virtual file system adapter for Azure Blob storage, used for remote cache.
-RUN curl -sSL -O https://packages.microsoft.com/config/ubuntu/22.04/packages-microsoft-prod.deb
-RUN dpkg -i packages-microsoft-prod.deb
-RUN rm packages-microsoft-prod.deb
-RUN apt-get update && \
-    apt-get install -y fuse3 blobfuse2
```
Comment (ScottTodd) [inline dockerfiles/cpubuilder_ubuntu_jammy.Dockerfile:30]: These changes are also in https://github.com/iree-org/base-docker-images/pull/10. The other changes here are harder to separate into their own smaller PRs.

File: dockerfiles/cpubuilder_ubuntu_jammy.Dockerfile [hunk 2]
```diff
@@ -15,32 +18,19 @@ RUN apt update && \
...
         libzstd-dev llvm-dev pkg-config
+
+# TODO: re-enable this when LLVM apt packages are working again
 # Recent compiler tools for build configurations like ASan/TSan.
 #   * See https://apt.llvm.org/ for context on the apt commands.
-ARG LLVM_VERSION=19
-RUN echo "deb http://apt.llvm.org/jammy/ llvm-toolchain-jammy-${LLVM_VERSION} main" >> /etc/apt/sources.list && \
-    curl https://apt.llvm.org/llvm-snapshot.gpg.key | gpg --dearmor > /etc/apt/trusted.gpg.d/llvm-snapshot.gpg && \
-    apt update && \
-    apt install -y clang-${LLVM_VERSION} lld-${LLVM_VERSION}
```
Comment (ScottTodd) [inline dockerfiles/cpubuilder_ubuntu_jammy.Dockerfile:24]: See https://github.com/llvm/llvm-project/issues/109276

**Comment by ScottTodd** [comment]
cc @MacDue @banach-space

**Comment by banach-space** [comment]
> cc @MacDue @banach-space 

Nice! I've skimmed through and all looks good to me. From what I can tell, the only platform-dependant part is the installation of `ccache`, right? I can also give this a spin tomorrow to double check this works on Aarch64 hosts  (pre-commit CI as a service 😅).

Btw, how can I add `qemu` locally?

**Comment by ScottTodd** [comment]
> From what I can tell, the only platform-dependant part is the installation of `ccache`, right?

Yep! The rest leans on already published multi-arch packages and portable scripts. Turns out multi-arch builds work well when they are single stage. We made things overcomplicated on https://github.com/iree-org/iree/pull/14372 with the multistage builds (`base-[platform]` --> `feature-a-[platform]` --> `feature-a-and-b-[platform]`)

> I can also give this a spin tomorrow to double check this works on Aarch64 hosts (pre-commit CI as a service 😅).

We still have an arm64 runner in the IREE repo for now (that's the only runner that Google is still providing), so I was planning on porting https://github.com/iree-org/iree/blob/main/.github/workflows/ci_linux_arm64_clang.yml after landing this and getting the first images published to GitHub's registry. Might need a bit of back and forth as I can't test the workflows and dockerfile changes in a single PR.

> Btw, how can I add `qemu` locally?

The install steps from 
https://github.com/iree-org/iree/blob/782f372b070eadd593a727004cf61dc84aabc634/build_tools/docker/dockerfiles/base-arm64.Dockerfile#L75-L80
should still work, but I want to see if https://github.com/docker/setup-qemu-action solves the same problem, so then we can stop mirroring that file. Instructions for how to generate that file were posted [here on Discord](https://discord.com/channels/689900678990135345/1166024193599615006/1196426431643783168):

> 
> Hey, an easy way I've found to build for AArch64 on x86_64, is to run the Dockerfile under QEMU (turtles all the way down :P).
> 
> First run:
> (Note, see: https://github.com/multiarch/qemu-user-static for more details, this sets up some binfmt_misc magic on your host system to allow it to run AArch64 (and other) binaries directly via qemu-*-user).
> 
> `docker run --rm --privileged multiarch/qemu-user-static:register`
> 
> 
> Then make a folder somewhere with this `Dockerfile`: https://gist.githubusercontent.com/MacDue/65e1351e2edfacc1a6c52f8b86e77a1d/raw/b9aeacc27149190e633358ca6b68b6a1349dfe46/Dockerfile
> Also, download the QEMU sources into that same folder via wget https://download.qemu.org/qemu-8.2.0.tar.xz (the download must be called qemu-8.2.0.tar.xz).
> 
> Finally, run: `DOCKER_BUILDKIT=1 docker build --network host --file Dockerfile --output out .`
> 
> Once the build completes `./out/qemu-aarch64` will contain an AArch64 binary for `qemu-aarch64`.

File: dockerfiles/cpubuilder_ubuntu_jammy.Dockerfile (hunk 2 above)
Comment (ScottTodd) [reply dockerfiles/cpubuilder_ubuntu_jammy.Dockerfile:24]: Also copied these llvm 19 changes to their own PR for easier review: https://github.com/iree-org/base-docker-images/pull/12

(renaming the file makes dependent PRs tricky though)

**Review by marbre — APPROVED** [review]
This look really nice!

File: .github/workflows/publish_cpubuilder.yml [hunk 3]
```diff
@@ -38,7 +38,8 @@ jobs:
         uses: docker/build-push-action@f2a1d5e99d037542a71f64918e516c093c6f3fc4
         with:
           context: .
-          file: dockerfiles/cpubuilder_ubuntu_jammy_x86_64.Dockerfile
+          file: dockerfiles/cpubuilder_ubuntu_jammy.Dockerfile
+          platforms: linux/amd64,linux/arm64
```
Comment (ScottTodd) [inline .github/workflows/publish_cpubuilder.yml:42]: https://github.com/iree-org/base-docker-images/actions/runs/10949036140/job/30401394373#step:5:135

```
 ERROR: Multi-platform build is not supported for the docker driver.
Switch to a different driver, or turn on the containerd image store, and try again.
Learn more at https://docs.docker.com/go/build-multi-platform/
Error: buildx failed with: Learn more at https://docs.docker.com/go/build-multi-platform/
```

Aww... I hit the same error locally while developing. The recommended steps in those docs worked for me, but I need to see what the docker/build-push-action needs.

Comment (ScottTodd) [reply .github/workflows/publish_cpubuilder.yml:42]: Fixed? https://github.com/iree-org/base-docker-images/commit/a61eb38120d704917e0b87855a9b106d224c91a8

So far so good: https://github.com/iree-org/base-docker-images/actions/runs/10949139831

Comment (ScottTodd) [reply .github/workflows/publish_cpubuilder.yml:42]: Yep, successful build and push. The workflow run time went up from 2 minutes to 20 minutes, in large part due to the ccache build from source.

Before: https://github.com/iree-org/base-docker-images/actions/runs/10948990985
After: https://github.com/iree-org/base-docker-images/actions/runs/10949139831

New package: https://github.com/iree-org/base-docker-images/pkgs/container/cpubuilder_ubuntu_jammy

**Comment by ScottTodd** [comment]
> > I can also give this a spin tomorrow to double check this works on Aarch64 hosts (pre-commit CI as a service 😅).
> 
> We still have an arm64 runner in the IREE repo for now (that's the only runner that Google is still providing), so I was planning on porting https://github.com/iree-org/iree/blob/main/.github/workflows/ci_linux_arm64_clang.yml after landing this and getting the first images published to GitHub's registry. Might need a bit of back and forth as I can't test the workflows and dockerfile changes in a single PR.

Starting to test at https://github.com/iree-org/iree/actions/runs/10949782485/job/30403753082 with this code: https://github.com/iree-org/iree/compare/users/scotttodd/ci-arm64?expand=1. setup-qemu-action did not do what I wanted :P

**Pull request merged**